### PR TITLE
[Ready to review] Spatial gradient 1d

### PR DESCRIPTION
**Pull request by ducha-aiki**
To be merged after or instead of https://github.com/arraiyopensource/kornia/pull/197
Added non-Sobel, but non-smoothed 1d (vertical and horizontal) version of spatial gradients - for future use with Harris and Hessin on spatial pyramid

Upd: refactored-out all the filter kernels into kornia.filters.kernels

### CODE REVIEW

File: kornia/filters/sobel.py [hunk 1]
```diff
@@ -12,6 +12,15 @@ def _get_sobel_kernel_3x3() -> torch.Tensor:
     ])
 
 
+def _get_separate_kernel_3x3() -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: should we place this to a kornia.filters.kernels.py file ?

File: kornia/filters/sobel.py [hunk 2]
```diff
@@ -28,16 +37,29 @@ class SpatialGradient(nn.Module):
         >>> output = kornia.filters.SpatialGradient()(input)  # 1x3x2x4x4
     """
 
-    def __init__(self) -> None:
+    def __init__(self, mode: str = 'Sobel') -> None:
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: I would use lower case -> `sobel` in order to be more consistent with pytorch standards

File: kornia/filters/sobel.py [hunk 3]
```diff
@@ -28,16 +37,29 @@ class SpatialGradient(nn.Module):
...
         super(SpatialGradient, self).__init__()
-        self.kernel: torch.Tensor = self.get_sobel_kernel()
+        if mode not in ['Sobel', 'NoBlur']:
+            raise TypeError("Mode should be either type is Sobel\
+                             or NoBlur. Got {}".format(mode))
+        self.mode = mode
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: `self.mode: str`

File: kornia/filters/sobel.py [hunk 4]
```diff
@@ -28,16 +37,29 @@ class SpatialGradient(nn.Module):
...
+        if self.mode == 'Sobel':
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: I don't like if else statements here, maybe we could have under filters.kernels a dict that returns the kernel given its key id

File: kornia/filters/sobel.py [hunk 5]
```diff
@@ -52,7 +74,10 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
 
         # convolve input tensor with sobel kernel
         kernel_flip: torch.Tensor = kernel.flip(-3)
-        return F.conv3d(input[:, :, None], kernel_flip, padding=1, groups=c)
+        # Pad with "replicate for spatial dims, but with zeros for channel dim
+        padded_inp = F.pad(F.pad(input, [1, 1, 1, 1], 'replicate')[:, :, None],
```
Comment (edgarriba) [inline kornia/filters/sobel.py]: `padded_inp: torch.Tensor`

File: kornia/filters/sobel.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/filters/sobel.py]: We probably should, but then I would move there all the existing - gaussian, box, sobel, which is  a separate refactoring effort

Comment (ducha-aiki) [reply kornia/filters/sobel.py]: I`ll do an issue for that

Comment (ducha-aiki) [reply kornia/filters/sobel.py]: Changed my mind - done in current PR

**Comment by ducha-aiki** [comment]
Looks like conda torch-nightly is down :(

**Comment by ducha-aiki** [comment]
Code coverage is down, because some of previously internal functions are exposed now.
